nighres/registration/embedded_syn.py

```diff
@@ -0,0 +1,235 @@
+        Dictionary collecting outputs under the following keys
+        (suffix of output files in brackets)
+
+        * deformed_source (niimg): Deformed source image (_esyn_def)
+        * mapping (niimg): Coordinate mapping from source to target (_esyn_map)
+        * inverse (niimg): Inverse coordinate mapping from target to source (_esyn_invmap) 
+
+    Notes
+    ----------
+    Port of the CBSTools Java module by Pierre-Louis Bazin. The main algorithm
+    is part of the ANTs software by Brian Avants and colleagues [1]_
+
+    References
+    ----------
+    .. [1] Avants BB, Epstein CL, Grossman M, Gee JC, Symmetric diffeomorphic 
+       image registration with cross-correlation: evaluating automated labeling 
+       of elderly and neurodegenerative brain, Med Image Anal. 2008 Feb;12(1):26-41
+    """
+
+    print('\nEmbedded SyN')
+
+    # make sure that saving related parameters are correct
+    if save_data:
+        output_dir = _output_dir_4saving(output_dir, source_image)
+
+        deformed_source_file = _fname_4saving(file_name=file_name,
+                                   rootfile=source_image,
+                                   suffix='esyn_def')
+
+        mapping_file = _fname_4saving(file_name=file_name,
+                                   rootfile=source_image,
+                                   suffix='esyn_map')
+
+        inverse_mapping_file = _fname_4saving(file_name=file_name,
+                                   rootfile=source_image,
+                                   suffix='esyn_invmap')
+
+     # load and get dimensions and resolution from input images
+    source = load_volume(source_image)
+    src_affine = source.affine
+    src_header = source.header
+    nsx = source.header.get_data_shape()[X]
+    nsy = source.header.get_data_shape()[Y]
+    nsz = source.header.get_data_shape()[Z]
+    rsx = source.header.get_zooms()[X]
+    rsy = source.header.get_zooms()[Y]
+    rsz = source.header.get_zooms()[Z]
+
+    target = load_volume(target_image)
+    trg_affine = target.affine
+    trg_header = target.header
+    ntx = target.header.get_data_shape()[X]
+    nty = target.header.get_data_shape()[Y]
+    ntz = target.header.get_data_shape()[Z]
+    rtx = target.header.get_zooms()[X]
+    rty = target.header.get_zooms()[Y]
+    rtz = target.header.get_zooms()[Z]
+
+    # build coordinate mapping matrices and save them to disk
+    src_coord = np.zeros((nsx,nsy,nsz,3))
+    trg_coord = np.zeros((ntx,nty,ntz,3))
+    for x in xrange(nsx):
+        for y in xrange(nsy):
+            for z in xrange(nsz):
+                src_coord[x,y,z,X] = x
+                src_coord[x,y,z,Y] = y
+                src_coord[x,y,z,Z] = z
+    src_map = nb.Nifti1Image(src_coord, source.affine, source.header)
+    src_map_file = _fname_4saving(file_name=file_name,
+                            rootfile=source_image,
+                            suffix='tmp_srccoord')
+    save_volume(os.path.join(output_dir, src_map_file), src_map)
```

Comment:
output_dir is an optional input, what if it is not given? maybe save to the tmp folder instead?

Reply:
actually, output_dir should always be set to something meaningful (if you don't give any, the default is to use the current folder, right? so io._output_dir_4saving should return that in case of None input. No?